libs/ngxtension/inject-signal-history/src/inject-signal-history.ts

```diff
@@ -0,0 +1,184 @@
+import {
+	computed,
+	Injector,
+	Signal,
+	signal,
+	WritableSignal,
+} from '@angular/core';
+import { explicitEffect } from 'ngxtension/explicit-effect';
+
+interface SignalHistoryRecord<T> {
+	value: T;
+	timestamp: number;
+}
+
+/**
+ * Creates a history record with the current timestamp.
+ * @param value The value to store in the history record.
+ * @returns A SignalHistoryRecord object.
+ */
+function createHistoryRecord<T>(value: T): SignalHistoryRecord<T> {
+	return { value, timestamp: Date.now() };
+}
+
+/**
+ * Enhances a writable signal with undo/redo history functionality.
+ *
+ * @param source The writable signal to track.
+ * @param options Configuration options for the history.
+ * @returns An object with `history`, `undo`, `redo`, `canUndo`, and `canRedo` properties.
+ */
+export function injectSignalHistory<T>(
+	source: WritableSignal<T>,
+	options?: {
+		/**
+		 * The maximum number of history records to store.
+		 * @default 50
+		 */
+		capacity?: number;
+
+		/**
+		 * The injector to use for the effect.
+		 * @default undefined
+		 */
+		injector?: Injector;
+	},
+): {
+	/**
+	 * The history of changes to the source signal.
+	 */
+	history: Signal<SignalHistoryRecord<T>[]>;
+	/**
+	 * Undo the last change to the source signal.
+	 */
+	undo: () => void;
+	/**
+	 * Redo the last undone change to the source signal.
+	 */
+	redo: () => void;
+	/**
+	 * Reset the history to the current state.
+	 */
+	reset: () => void;
+	/**
+	 * Clear the history. This will remove all history records.
+	 */
+	clear: () => void;
+	/**
+	 * A signal indicating if undo is possible.
+	 */
+	canUndo: Signal<boolean>;
+	/**
+	 * A signal indicating if redo is possible.
+	 */
+	canRedo: Signal<boolean>;
+} {
+	const capacity = options?.capacity ?? 100; // Default capacity is 100 records
```

Comment:
I'd guard that function with `options?.injector && assertInInjectionContext(...);`

Reply:
Forgot about it!